src/PreserveModuleNamePlugin.ts

```diff
@@ -72,90 +164,320 @@ function getPreservedModules(modules: Webpack.Module[]) {
   return new Set(
     modules.filter(m => {
       // Some modules might have [preserveModuleName] already set, see ConventionDependenciesPlugin.
-      let value = m[preserveModuleName];      
-      for (let r of m.reasons) {
-        if (!r.dependency || !r.dependency[preserveModuleName]) continue;
-        value = true;
-        let req = removeLoaders((r.dependency as ModuleDependency).request);
-        // We try to find an absolute string and set that as the module [preserveModuleName], as it's the best id.
-        if (req && !req.startsWith(".")) {
-          m[preserveModuleName] = req;
-          return true;
-        }
+      if (m[preserveModuleName]) {
+        return true;
       }
-      return !!value;
+
+      // Preserve the module if its dependencies are also preserved
+      const reasons = (m.reasons && Array.isArray(m.reasons)) ? m.reasons : [];
```

Comment:
Is there a reason for this defensive code?
It was not in my code and I've never had an issue filed about `m.reasons` not being an array. 
Is there a case where it happens?

Reply:
I am unsure, I usually add safety checks and assume to worst. If you need me to I can remove it.

Reply:
The problem I have with defensive code is that it creates headache for future maintainers.

When I see this code it tells me "`m.reasons` could be something else than an array, we better be careful" and I immediately ask myself: "how so? what am I missing?".

If this `m.reasons` _can_ be `null` we should document how / why, it's useful information.
If it cannot, then it's misleading bloat that I'd rather remove.

When unsure, someone needs to figure this out and it's courteous that you do it rather than deferring the analysis to future maintainers.